pennylane/templates/layers/particle_conserving_u1.py

```diff
@@ -0,0 +1,267 @@
+
+    |
+
+    :math:`U_A(\phi)` is the unitary matrix
+
+    .. math::
+
+        U_A(\phi) = \left(\begin{array}{cc} 0 & e^{-i\phi} \\ e^{-i\phi} & 0 \\ \end{array}\right)
+
+    acting on the state of qubit ``m`` which is further decomposed in terms of the
+    `quantum operations <https://pennylane.readthedocs.io/en/stable/introduction/operations.html>`_
+    supported by preparesnnyLane,
+
+    |
+
+    .. figure:: ../../_static/templates/layers/ua_decomposition.png
+        :align: center
+        :width: 70%
+        :target: javascript:void(0);
+
+    |
+
+    where,
+
+    |
+
+    .. figure:: ../../_static/templates/layers/phaseshift_decomposition.png
+        :align: center
+        :width: 70%
+        :target: javascript:void(0);
+
+    |
+
+    The quantum circuits above decomposing the unitaries :math:`U_{1,\mathrm{ex}}(\phi, \theta)` and
+    :math:`U_A(\phi)` are implemented by the :func:`~.u1_ex_gate` and :func:`~.decompose_ua`
+    functions, respectively.
+
+    Args:
+        weights (array[float]): Array of weights of shape ``(D, M, 2)``.
+            ``D`` is the number of entangler block layers and ``M`` = :math:`N_\mathrm{qubits}-1`
+            is the number of exchange gates :math:`U_{1,\mathrm{ex}}` per layer.
+        wires (Iterable or Wires): Wires that the template acts on. Accepts an iterable of numbers
+            or strings, or a Wires object.
+        init_state (array[int]): Length ``len(wires)`` occupation-number vector representing the
+            HF state. ``init_state`` is used to initialize the wires.
+
+    Raises:
+        ValueError: if inputs do not have the correct format
+
+    .. UsageDetails::
+
+        Notice that:
+
+        #. The number of wires has to be equal to the number of spin orbitals included in
+           the active space.
+
+        #. The number of trainable parameters scales linearly with the number of layers as
+           :math:`2D(N_\mathrm{qubits}-1)`.
+
+        An example of how to use this template is shown below:
+
+        .. code-block:: python
+
+            import pennylane as qml
+            from pennylane.templates import ParticleConservingU1
+            from functools import partial
+
+            # Build the electronic Hamiltonian
+            h, qubits = qml.qchem.molecular_hamiltonian("h2", "h2.xyz")
+
+            # Define the HF state
+            ref_state = qml.qchem.hf_state(electrons=2, qubits)
+
+            # Define the device
+            dev = qml.device('default.qubit', wires=qubits)
+
+            # Define the ansatz
+            ansatz = partial(ParticleConservingU1, init_state=ref_state)
+
+            # Define the cost function
+            cost_fn = qml.VQECost(ansatz, h, dev)
+
+            # Compute the expectation value of 'h'
+            layers = 2
+            params = qml.init.particle_conserving_u1_normal(layers, qubits)
+            print(cost_fn(params))
+    """
+
+    wires = Wires(wires)
+
+    layers = weights.shape[0]
```

Comment:
As I mentioned, in tape mode ``weights`` could be anything like a numpy array, pennylane-numpy array, tf tensor, tf Variable etc. I think for now this will work though. 

We are soon getting a much nicer way to extract shapes and do classical processing :)

Reply:
Thanks for the note here.